### DISCUSSION TITLE
SimplePagination, no data from useQuery

### DISCUSSION THREAD

**Discussion by nephix** [urql-graphql/urql#2468] [Q&A] [answered]
Hey guys,

I'm trying to get the [`simplePagination`](https://formidable.com/open-source/urql/docs/graphcache/local-resolvers/#simple-pagination) to work in a project. I create a client like that:

```typescript
const cache = cacheExchange<GraphCacheConfig>({
  resolvers: {
    Query: {
      getAdmins: simplePagination(),
    },
  },
});

const client = createClient({
  url: API_URL,
  fetchOptions: {
    credentials: 'include',
  },
  exchanges: [devtoolsExchange, dedupExchange, cache, fetchExchange],
});
```

`GraphCacheConfig` is a type generated by GCG.

When I run the following query:

```typescript
const GET_ADMINS = gql`
  query GetAdmins($skip: Int!, $limit: Int!) {
    getAdmins(input: { skip: $skip, limit: $limit }) {
      ...AdminPart
    }
  }

  ${ADMIN_FRAGMENT}
`;

useQuery<GetAdminsQuery, GetAdminsQueryVariables>({
  variables,
  query: GET_ADMINS,
});
```

The schema for that query is:

```graphql
input GetAdminsInput {
  skip: Int!
  limit: Int = 20
}

type Query {
  getAdmins(input: GetAdminsInput!): [Admin!]
}
```

I can see in the network tab that an array of admin users is returned by the server:

![SCR-20220529-oj7](https://user-images.githubusercontent.com/3169205/170881282-3fe51dae-2a61-4fc8-ba74-e43ae0094f74.png)

However, in urql devtools I don't see data:

![SCR-20220529-ojc](https://user-images.githubusercontent.com/3169205/170881313-bc23705a-5f41-44e6-8703-21a8b7680f05.png)

as well as in the component that uses this query.

When I remove the resolver from the `cacheExchange` I get the data via `useQuery`, but can't use pagination and caching (I know I could do something like [here](https://github.com/FormidableLabs/urql/blob/main/examples/with-pagination/src/PaginatedNpmSearch.jsx) but would like to avoid that)

Any idea what the issue could be?

Cheers

**Comment by kitten** [comment, marked as answer]
The utility only works out of the box your schema matches the pagination scheme exactly. The top-level arguments may be renamed but they're expected to not be on an input object. We can't implement every form of pagination — hence Relay pagination exists as a spec — so this works only for the expected shape. 

https://formidable.com/open-source/urql/docs/api/graphcache/#simplepagination
The top-level arguments are renameable though.

**Reply by nephix** [reply]
Makes sense, thanks

**Comment by nandorojo** [comment]
I'm facing the same issue, even though my arguments are indeed top-level.

```ts
{
  myBookings: simplePagination({
    limitArgument: 'limit',
    offsetArgument: 'offset',
    mergeMode: 'after'
  })
}
```

This is working fine if I add no variables to my query other than `limit` and `offset`. However, once I add other variables, urql always returns an empty list for `myBookings(limit, offset, sort)`.

**Reply by nephix** [reply]
It works now for me after I moved the arguments to the top level, with additional params as well:

```graphql
getUsers(skip: $skip, limit: $limit, searchFilter: $searchFilter) {
   ...
}
```

and

```typescript
const cache = cacheExchange<GraphCacheConfig>({
  resolvers: {
    Query: {
      getUsers: simplePagination(),
    },
  }
});
```

**Reply by nandorojo** [reply]
Interesting, thanks for the heads up. I’ll see if I can reproduce it.
